src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/CQLSessionCache.java

```diff
@@ -0,0 +1,200 @@
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import java.net.InetSocketAddress;
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * CQL session cache to reuse the session for the same tenant and token. The cache is configured to
+ * expire after <code>CACHE_TTL_SECONDS</code> of inactivity and to have a maximum size of <code>
+ * CACHE_TTL_SECONDS</code> sessions.
+ */
+@ApplicationScoped
+public class CQLSessionCache {
+  private static final Logger LOGGER = LoggerFactory.getLogger(JsonApiStartUp.class);
+
+  /** Configuration for the JSON API operations. */
+  private final OperationsConfig operationsConfig;
+
+  /** Stargate request info. */
+  @Inject StargateRequestInfo stargateRequestInfo;
+
+  /**
+   * Default tenant to be used when the backend is OSS cassandra and when no tenant is passed in the
+   * request
+   */
+  private static final String DEFAULT_TENANT = "default_tenant";
+  /** CQL username to be used when the backend is AstraDB */
+  private static final String TOKEN = "token";
+
+  /** CQLSession cache. */
+  private final Cache<SessionCacheKey, CqlSession> sessionCache;
+
+  public static final String ASTRA = "astra";
+  public static final String CASSANDRA = "cassandra";
+  /** Default token property name which will be used by the integration tests */
+  @Inject
+  public CQLSessionCache(OperationsConfig operationsConfig) {
+    this.operationsConfig = operationsConfig;
+    sessionCache =
+        Caffeine.newBuilder()
+            .expireAfterAccess(
+                Duration.ofSeconds(operationsConfig.databaseConfig().sessionCacheTtlSeconds()))
+            .maximumSize(operationsConfig.databaseConfig().sessionCacheMaxSize())
+            .evictionListener(
+                (RemovalListener<SessionCacheKey, CqlSession>)
+                    (sessionCacheKey, session, cause) -> {
+                      if (sessionCacheKey != null) {
+                        if (LOGGER.isDebugEnabled()) {
+                          LOGGER.debug(
+                              "Removing session for tenant : {}", sessionCacheKey.tenantId);
+                        }
+                      }
+                      if (session != null) {
+                        session.close();
+                      }
+                    })
+            .build();
+    LOGGER.info(
+        "CQLSessionCache initialized with ttl of {} seconds and max size of {}",
+        operationsConfig.databaseConfig().sessionCacheTtlSeconds(),
+        operationsConfig.databaseConfig().sessionCacheMaxSize());
+  }
+
+  /**
+   * Loader for new CQLSession.
+   *
+   * @return CQLSession
+   * @throws RuntimeException if database type is not supported
+   */
+  private CqlSession getNewSession(SessionCacheKey cacheKey) {
+    if (LOGGER.isDebugEnabled()) {
```

Comment:
Can we change some of the debug log to trace?

Reply:
Fixed in this commit 9a6b2efd530b38e4d17e64bad3cad014cc3dbccb